packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright 2022 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaLabelingProps} from '@react-types/shared';
+import {HTMLAttributes, MutableRefObject, useCallback, useEffect, useState} from 'react';
+import {useLayoutEffect} from '@react-aria/utils';
+
+export type AriaLandmarkRole = 'main' | 'region' | 'search' | 'navigation' | 'form' | 'banner' | 'contentinfo' | 'complementary';
+
+export interface AriaLandmarkProps extends AriaLabelingProps {
+  role: AriaLandmarkRole
+}
+
+interface LandmarkAria {
+  landmarkProps: HTMLAttributes<HTMLElement>
+}
+
+type Landmark = {
+  ref: MutableRefObject<HTMLElement>,
+  role: AriaLandmarkRole,
+  label?: string,
+  lastFocused?: HTMLElement,
+  onFocus: () => void,
+  onBlur: () => void
+};
+
+class LandmarkManager {
+  private landmarks: Array<Landmark> = [];
+  private static instance: LandmarkManager;
+
+  private constructor() {}
+
+  public static getInstance(): LandmarkManager {
+    if (!LandmarkManager.instance) {
+      LandmarkManager.instance = new LandmarkManager();
+      LandmarkManager.instance.setup();
+    }
+
+    return LandmarkManager.instance;
+  }
+
+  private setup() {
+    document.addEventListener('keydown', LandmarkManager.getInstance().f6Handler.bind(LandmarkManager.getInstance()), {capture: true});
+    document.addEventListener('focusin', LandmarkManager.getInstance().focusinHandler.bind(LandmarkManager.getInstance()), {capture: true});
+  }
+
+  private focusLandmark(landmark: HTMLElement) {
+    this.landmarks.find(l => l.ref.current === landmark)?.onFocus();
+  }
+
+  /**
+   * Return set of landmarks with a specific role.
+   */
+  public getLandmarksByRole(role: AriaLandmarkRole) {
+    return new Set(this.landmarks.filter(l => l.role === role));
+  }
+
+  /**
+   * Return first landmark with a specific role.
+   */
+  public getLandmarkByRole(role: AriaLandmarkRole) {
+    let landmarks = this.landmarks.filter(l => l.role === role);
+    if (landmarks.length > 0) {
+      return landmarks[0];
+    }
+  }
+
+  public addLandmark(newLandmark: Landmark) {
+    if (this.landmarks.find(landmark => landmark.ref === newLandmark.ref)) {
+      return;
+    }
+
+    if (this.landmarks.filter(landmark => landmark.role === 'main').length > 1) {
+      console.error('Page can contain no more than one landmark with the role "main".');
```

Comment:
followup: log the two elements with the main role so people can more easily find them.